File: frontend/src/api/career_simulation.js
import api from './base';

class CareerSimulationAPI {
    // Create a new career simulation
    async createSimulation(simulationRequest) {
        const response = await api.post('/career-simulation/simulate', simulationRequest);
        return response.data;
    }

    // Get a specific simulation by ID
    async getSimulation(simulationId) {
        const response = await api.get(`/career-simulation/${simulationId}`);
        return response.data;
    }

    // Get all simulations for a specific offer
    async getSimulationsForOffer(offerId) {
        const response = await api.get(`/career-simulation/offer/${offerId}`);
        return response.data;
    }

    // Get all simulations for the authenticated user
    async getUserSimulations() {
        const response = await api.get('/career-simulation/me');
        return response.data;
    }

    // Delete a simulation
    async deleteSimulation(simulationId) {
        const response = await api.delete(`/career-simulation/${simulationId}`);
        return response.data;
    }

    async compareOffers(payload) {
        const response = await api.post('/career-simulation/compare', payload);
        return response.data;
    }

    // Get simulation statistics
    async getSimulationStatistics() {
        const response = await api.get('/career-simulation/statistics/me');
        return response.data;
    }

    // Helper method to create a simulation request
    createSimulationRequest(offerId, options = {}) {
        const isBlank = (v) => v === "" || v === null || v === undefined;
        const numOrUndefined = (v) => {
            if (isBlank(v)) return undefined;
            const n = typeof v === "number" ? v : parseFloat(String(v));
            return Number.isFinite(n) ? n : undefined;
        };
        const intOrUndefined = (v) => {
            if (isBlank(v)) return undefined;
            const n = typeof v === "number" ? v : parseInt(String(v), 10);
            return Number.isFinite(n) ? n : undefined;
        };
        const bonusOrUndefined = (v) => {
            if (isBlank(v)) return undefined;
            if (typeof v === "number") return Number.isFinite(v) ? v : undefined;
            const s = String(v).trim();
            if (!s) return undefined;
            const n = Number(s.replace(/,/g, ""));
            if (Number.isFinite(n) && !/%|\$|k$/i.test(s)) {
                return n;
            }
            return s;
        };

        const milestones = (options.milestones || []).map((m) => {
            const year = intOrUndefined(m?.year) ?? 1;
            return {
                year,
                title: isBlank(m?.title) ? undefined : m.title,
                raise_percent: numOrUndefined(m?.raise_percent),
                new_base_salary: numOrUndefined(m?.new_base_salary),
                bonus_expected: numOrUndefined(m?.bonus_expected),
                equity_value: numOrUndefined(m?.equity_value),
            };
        });

        const req = {
            offer_id: offerId,
            simulation_years: options.simulationYears ?? 5,
            success_criteria: options.successCriteria ?? options.success_criteria ?? [
                {
                    criteria_type: 'salary',
                    weight: 0.4,
                    target_value: 150000,
                    importance: 'high',
                    description: 'Competitive salary growth'
                },
                {
                    criteria_type: 'work_life_balance',
                    weight: 0.3,
                    target_value: 8,
                    importance: 'high',
                    description: 'Good work-life balance'
                },
                {
                    criteria_type: 'learning_opportunities',
                    weight: 0.2,
                    target_value: 8,
                    importance: 'medium',
                    description: 'Continuous learning and growth'
                },
                {
                    criteria_type: 'impact',
                    weight: 0.1,
                    target_value: 7,
                    importance: 'medium',
                    description: 'Meaningful work and impact'
                }
            ],
        };

        req.personal_growth_rate = numOrUndefined(options.personalGrowthRate) ?? 0.5;
        req.risk_tolerance = numOrUndefined(options.riskTolerance) ?? 0.5;
        req.job_change_frequency = numOrUndefined(options.jobChangeFrequency) ?? 2.5;
        req.geographic_flexibility = options.geographicFlexibility ?? true;
        req.industry_switch_willingness = options.industrySwitchWillingness ?? false;
        req.inflation_rate = numOrUndefined(options.inflationRate) ?? 0.025;
        req.market_growth_rate = numOrUndefined(options.marketGrowthRate) ?? 0.05;
        if (!isBlank(options.industryTrendOverride)) req.industry_trend_override = options.industryTrendOverride;

        const startingSalary = numOrUndefined(options.startingSalary ?? options.starting_salary);
        if (startingSalary !== undefined) req.starting_salary = startingSalary;

        const annualRaisePercent = numOrUndefined(options.annualRaisePercent ?? options.annual_raise_percent);
        if (annualRaisePercent !== undefined) req.annual_raise_percent = annualRaisePercent;

        const raiseScenarios = options.raiseScenarios ?? options.raise_scenarios;
        if (raiseScenarios && typeof raiseScenarios === "object") {
            const normalized = {
                conservative: numOrUndefined(raiseScenarios.conservative),
                expected: numOrUndefined(raiseScenarios.expected),
                optimistic: numOrUndefined(raiseScenarios.optimistic),
            };
            if (Object.values(normalized).some((v) => v !== undefined)) {
                req.raise_scenarios = normalized;
            }
        }

        if (milestones.length > 0) req.milestones = milestones;

        const annualBonus = bonusOrUndefined(options.annualBonus ?? options.annual_bonus);
        if (annualBonus !== undefined) req.annual_bonus = annualBonus;

        const annualEquity = numOrUndefined(options.annualEquity ?? options.annual_equity);
        if (annualEquity !== undefined) req.annual_equity = annualEquity;

        if (!isBlank(options.notes)) req.notes = options.notes;

        return req;
    }

    // Helper method to format simulation response for display
    formatSimulationForDisplay(simulation) {
        if (!simulation || !simulation.response) {
            return null;
        }

        const response = simulation.response;
        return {
            id: simulation.simulation_id,
            status: simulation.status,
            createdAt: simulation.created_at,
            completedAt: simulation.updated_at,
            computationTime: simulation.computation_time_seconds,
            
            // Career paths
            careerPaths: response.career_paths || [],
            optimalPath: response.optimal_path,
            pathRankings: response.path_rankings || [],
            
            // Insights
            decisionInsights: response.decision_insights || [],
            riskAssessments: response.risk_assessments || [],
            opportunityAnalysis: response.opportunity_analysis || [],
            
            // Recommendations
            nextStepRecommendation: response.next_step_recommendation,
            longTermStrategy: response.long_term_strategy,
            
            // Metadata
            confidenceLevel: response.confidence_level,
            dataSources: response.data_sources || [],
            
            // Request details
            request: simulation.request
        };
    }

    // Helper method to extract key metrics from simulation
    extractKeyMetrics(simulation) {
        if (!simulation || !simulation.response || !simulation.response.optimal_path) {
            return null;
        }

        const optimalPath = simulation.response.optimal_path;
        
        return {
            totalEarnings5yr: optimalPath.total_earnings_5yr,
            totalEarnings10yr: optimalPath.total_earnings_10yr,
            peakSalary: optimalPath.peak_salary,
            careerGrowthRate: optimalPath.career_growth_rate,
            overallScore: optimalPath.overall_score,
            
            // Component scores
            salaryScore: optimalPath.salary_score,
            workLifeBalanceScore: optimalPath.work_life_balance_score,
            learningScore: optimalPath.learning_score,
            impactScore: optimalPath.impact_score,
            
            // Career progression
            titleProgression: optimalPath.title_progression || [],
            companiesWorked: optimalPath.companies_worked || [],
            
            // Probability outcomes
            probabilityOutcomes: optimalPath.probability_outcomes || [],
            
            // Decision points
            decisionPoints: optimalPath.decision_points || []
        };
    }

    // Helper method to compare multiple simulations
    compareSimulations(simulations) {
        if (!simulations || simulations.length === 0) {
            return [];
        }

        return simulations.map(sim => {
            const metrics = this.extractKeyMetrics(sim);
            if (!metrics) return null;

            return {
                simulationId: sim.simulation_id,
                offerId: sim.request.offer_id,
                status: sim.status,
                completedAt: sim.updated_at,
                ...metrics,
                // Add comparative metrics
                earningsPerYear5yr: metrics.totalEarnings5yr / 5,
                earningsPerYear10yr: metrics.totalEarnings10yr / 10,
                growthPotential: metrics.careerGrowthRate * 100
            };
        }).filter(Boolean);
    }

    // Helper method to get simulation progress
    getSimulationProgress(simulation) {
        if (!simulation) return null;

        switch (simulation.status) {
            case 'pending':
                return { status: 'pending', message: 'Simulation is queued', progress: 0 };
            case 'running':
                return { status: 'running', message: 'Simulation is running', progress: 50 };
            case 'completed':
                return { status: 'completed', message: 'Simulation completed', progress: 100 };
            case 'failed':
                return { 
                    status: 'failed', 
                    message: simulation.error_message || 'Simulation failed', 
                    progress: 0 
                };
            default:
                return { status: 'unknown', message: 'Unknown status', progress: 0 };
        }
    }
}

const careerSimulationAPI = new CareerSimulationAPI();
export default careerSimulationAPI;

File: backend/schema/CareerSimulation.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any, Union
from enum import Enum

class IndustryTrend(str, Enum):
    GROWING = "growing"
    STABLE = "stable"
    DECLINING = "declining"
    VOLATILE = "volatile"

class CompanyStage(str, Enum):
    STARTUP = "startup"
    EARLY_STAGE = "early_stage"
    GROWTH_STAGE = "growth_stage"
    MATURE = "mature"
    ENTERPRISE = "enterprise"

class EconomicCondition(str, Enum):
    BOOM = "boom"
    EXPANSION = "expansion"
    STABLE = "stable"
    CONTRACTION = "contraction"
    RECESSION = "recession"

class SuccessCriteriaType(str, Enum):
    SALARY = "salary"
    WORK_LIFE_BALANCE = "work_life_balance"
    LEARNING_OPPORTUNITIES = "learning_opportunities"
    IMPACT = "impact"
    CAREER_GROWTH = "career_growth"
    JOB_SECURITY = "job_security"

class CareerMilestone(BaseModel):
    """Career milestone in simulation timeline"""
    year: int
    title: str
    company: str
    salary_base: float
    salary_total: float
    equity_value: float = 0
    bonus_expected: float = 0
    responsibilities_level: int  # 1-10 scale
    learning_score: int  # 1-10 scale
    work_life_balance: int  # 1-10 scale
    impact_score: int  # 1-10 scale
    job_security_score: int  # 1-10 scale
    market_demand: int  # 1-10 scale
    career_satisfaction: int  # 1-10 scale

class ProbabilityOutcome(BaseModel):
    """Probability distribution for career outcomes"""
    scenario: str  # best_case, average_case, worst_case
    probability: float  # 0.0 to 1.0
    final_salary_5yr: float
    final_salary_10yr: float
    total_earnings_5yr: float
    total_earnings_10yr: float
    final_title: str
    final_company_stage: CompanyStage
    career_satisfaction: int
    work_life_balance: int
    learning_achievements: List[str]

class DecisionPoint(BaseModel):
    """Key decision points where career paths diverge"""
    year: int
    decision_type: str  # promotion, job_change, company_switch, industry_pivot
    options: List[str]
    impact_analysis: Dict[str, Any]
    recommendation: str
    confidence_level: float  # 0.0 to 1.0

class CareerPath(BaseModel):
    """Complete career path simulation"""
    path_id: str
    path_name: str
    base_offer_id: str  # Reference to the starting offer
    simulation_years: int  # 5 or 10
    
    # Input Parameters
    industry_trend: IndustryTrend
    company_stage: CompanyStage
    economic_condition: EconomicCondition
    personal_growth_rate: float  # 0.0 to 1.0
    risk_tolerance: float  # 0.0 to 1.0
    job_change_frequency: float  # Average years between job changes
    
    # Career Trajectory
    milestones: List[CareerMilestone]
    
    # Outcomes
    total_earnings_5yr: float
    total_earnings_10yr: float
    peak_salary: float
    career_growth_rate: float  # Annual percentage
    title_progression: List[str]
    companies_worked: List[str]
    
    # Probability Analysis
    probability_outcomes: List[ProbabilityOutcome]
    
    # Decision Points
    decision_points: List[DecisionPoint]
    
    # Success Metrics
    overall_score: float  # 0-100
    salary_score: float  # 0-100
    work_life_balance_score: float  # 0-100
    learning_score: float  # 0-100
    impact_score: float  # 0-100

class SuccessCriteria(BaseModel):
    """User-defined success criteria for career evaluation"""
    criteria_type: SuccessCriteriaType
    weight: float  # 0.0 to 1.0
    target_value: Union[float, int, str]
    importance: str  # critical, high, medium, low
    description: str


class UserMilestone(BaseModel):
    year: int
    title: Optional[str] = None
    raise_percent: Optional[float] = None
    new_base_salary: Optional[float] = None
    bonus_expected: Optional[float] = None
    equity_value: Optional[float] = None

class CareerSimulationRequest(BaseModel):
    """Request to run career simulation"""
    offer_id: str
    simulation_years: int = 5  # 5 or 10
    success_criteria: List[SuccessCriteria]
    
    # Personal Factors
    personal_growth_rate: float = 0.5  # 0.0 to 1.0
    risk_tolerance: float = 0.5  # 0.0 to 1.0
    job_change_frequency: float = 2.5  # Average years between changes
    geographic_flexibility: bool = True
    industry_switch_willingness: bool = False
    
    # Economic Assumptions
    inflation_rate: float = 0.025  # 2.5% annual
    market_growth_rate: float = 0.05  # 5% annual
    industry_trend_override: Optional[IndustryTrend] = None

    starting_salary: Optional[float] = None
    annual_raise_percent: float = 3.0
    raise_scenarios: Optional[Dict[str, float]] = None
    milestones: Optional[List[UserMilestone]] = None
    annual_bonus: Optional[float] = None
    annual_equity: Optional[float] = None
    notes: Optional[str] = None

class CareerSimulationResponse(BaseModel):
    """Response containing career simulation results"""
    simulation_id: str
    request: CareerSimulationRequest
    
    # Multiple Career Paths
    career_paths: List[CareerPath]
    
    # Comparative Analysis
    optimal_path: CareerPath
    path_rankings: List[Dict[str, Any]]
    
    # Key Insights
    decision_insights: List[str]
    risk_assessments: List[str]
    opportunity_analysis: List[str]
    
    # Recommendations
    next_step_recommendation: str
    long_term_strategy: str
    
    # Metadata
    generated_at: str
    confidence_level: float  # Overall confidence in predictions
    data_sources: List[str]

class SimulationParameters(BaseModel):
    """Internal parameters for simulation engine"""
    # Salary Growth Factors
    base_salary_growth_annual: float = 0.03  # 3% base
    promotion_salary_boost: float = 0.15  # 15% on promotion
    job_change_salary_boost: float = 0.12  # 12% on job change
    
    # Title Progression Timeline (years)
    junior_to_mid: float = 2.0
    mid_to_senior: float = 3.0
    senior_to_lead: float = 4.0
    lead_to_principal: float = 5.0
    principal_to_staff: float = 6.0
    
    # Company Stage Multipliers
    startup_growth_multiplier: float = 1.4
    early_stage_growth_multiplier: float = 1.2
    growth_stage_growth_multiplier: float = 1.1
    mature_growth_multiplier: float = 0.8
    enterprise_growth_multiplier: float = 0.6
    
    # Industry Trend Multipliers
    growing_industry_multiplier: float = 1.3
    stable_industry_multiplier: float = 1.0
    declining_industry_multiplier: float = 0.7
    volatile_industry_multiplier: float = 1.1
    
    # Economic Condition Multipliers
    boom_economy_multiplier: float = 1.2
    expansion_economy_multiplier: float = 1.1
    stable_economy_multiplier: float = 1.0
    contraction_economy_multiplier: float = 0.8
    recession_economy_multiplier: float = 0.6

class CareerSimulation(BaseModel):
    """Complete career simulation document"""
    # Metadata
    user_uuid: str
    simulation_id: str
    created_at: str
    updated_at: str
    
    # Request and Response
    request: CareerSimulationRequest
    response: CareerSimulationResponse
    
    # Simulation Parameters Used
    parameters: SimulationParameters
    
    # Status
    status: str  # pending, running, completed, failed
    error_message: Optional[str] = None
    
    # Analytics
    computation_time_seconds: Optional[float] = None
    scenarios_generated: int = 0
    confidence_score: Optional[float] = None
